rmf_robot_sim_gz_plugins/src/TeleportDispenser.cpp

```diff
@@ -101,6 +127,186 @@ TeleportDispenserPlugin::~TeleportDispenserPlugin()
+        send_dispenser_response(DispenserResult::FAILED);
+      }
+      return;
+    }
+
+    _dispense = true; // Mark true to dispense item next time PreUpdate() is called
+  }
+}
+
+void TeleportDispenserPlugin::on_update(EntityComponentManager& ecm)
+{
+  try_refill_dispenser(ecm);
+
+  // periodic pub on dispenser state
+  constexpr double interval = 2.0;
+  if (_sim_time - _last_pub_time >= interval || _dispense)
+  {
+    _last_pub_time = _sim_time;
+    _current_state.time = simulation_now(_sim_time);
+
+    if (_dispense)
+    {
+      _current_state.mode = DispenserState::BUSY;
+      _current_state.request_guid_queue = {_latest.request_guid};
+    }
+    else
+    {
+      _current_state.mode = DispenserState::IDLE;
+      _current_state.request_guid_queue.clear();
+    }
+    _state_pub->publish(_current_state);
+  }
+
+  // `dispense` is set to true if the dispenser plugin node has
+  // received a valid DispenserRequest
+  if (_dispense)
+  {
+    send_dispenser_response(DispenserResult::ACKNOWLEDGED);
+
+    bool is_success = false;
+    if (_dispenser_filled)
+    {
+      RCLCPP_INFO(_ros_node->get_logger(), "Dispensing item");
+      bool res = dispense_on_nearest_robot(ecm, _latest.transporter_type);
+      if (res)
+      {
+        is_success = true;
+        send_dispenser_response(DispenserResult::SUCCESS);
+        RCLCPP_INFO(_ros_node->get_logger(), "Success");
+      }
+      else
+      {
+        send_dispenser_response(DispenserResult::FAILED);
+        RCLCPP_WARN(_ros_node->get_logger(), "Unable to dispense item");
+      }
+    }
+    else
+    {
+      RCLCPP_WARN(_ros_node->get_logger(),
+        "No item to dispense: [%s]", _latest.request_guid.c_str());
+      send_dispenser_response(DispenserResult::FAILED);
+    }
+
+    _past_request_guids.emplace(_latest.request_guid, is_success);
+
+    _dispense = false;
+  }
+}
+
+void TeleportDispenserPlugin::try_refill_dispenser(EntityComponentManager& ecm)
+{
+  constexpr double interval = 2.0;
+  if (_sim_time - _last_pub_time >= interval)
+  {
+    // Occasionally check to see if dispensed item has been returned to it
+    if (!_dispenser_filled && _item_en_found &&
+      _dispenser_vicinity_box.Contains(
+        ecm.Component<components::Pose>(_item_en)->Data().Pos()))
+    {
+      _dispenser_filled = true;
+    }
+  }
+}
+
+bool TeleportDispenserPlugin::dispense_on_nearest_robot(
+  EntityComponentManager& ecm,
+  const std::string& fleet_name)
+{
+  if (!_dispenser_filled)
+    return false;
+
+  const auto fleet_state_it = _fleet_states.find(fleet_name);
+  if (fleet_state_it == _fleet_states.end())
+  {
+    RCLCPP_WARN(_ros_node->get_logger(),
+      "No such fleet: [%s]", fleet_name.c_str());
+    return false;
+  }
+
+  std::vector<SimEntity> robot_list;
+  fill_robot_list(ecm, fleet_state_it, robot_list);
+
+  bool found = false;
+  SimEntity robot_model = find_nearest_model(ecm, robot_list, found);
```

Comment:
Instead of a bool + entity, this function can just return a `std::optional<Entity>` and we can change the snippet below to check if the `optional` has a value